rs-matter/src/respond.rs

```diff
@@ -0,0 +1,269 @@
+    where
+        T: ExchangeHandler,
+        Self: Sized,
+    {
+        CompositeExchangeHandler(self, other)
+    }
+}
+
+impl<T> ExchangeHandler for &T
+where
+    T: ExchangeHandler,
+{
+    async fn handle(&self, exchange: &mut Exchange<'_>) -> Result<(), Error> {
+        (*self).handle(exchange).await
+    }
+}
+
+/// A struct for composing two exchange handlers into a single one, where each handler is handling one specific protocol (i.e. SC vs IM).
+pub struct CompositeExchangeHandler<F, S>(pub F, pub S);
+
+impl<F, S> ExchangeHandler for CompositeExchangeHandler<F, S>
+where
+    F: ExchangeHandler,
+    S: ExchangeHandler,
+{
+    async fn handle(&self, exchange: &mut Exchange<'_>) -> Result<(), Error> {
+        let result = self.0.handle(exchange).await;
+
+        match result {
+            Err(e) if e.code() == ErrorCode::InvalidProto => self.1.handle(exchange).await,
+            other => other,
+        }
+    }
+}
+
+/// A generic responder utility for accepting and handling exchanges received by the provided `Matter` stack,
+/// by applying the provided `ExchangeHandler` instance to each accepted exchange.
+///
+/// This responder uses an intra-task concurrency model - without an external executor - where all handling is done as a single future.
+pub struct Responder<'a, T> {
+    name: &'a str,
+    handler: T,
+    matter: &'a Matter<'a>,
+    respond_after_ms: u64,
+}
+
+impl<'a, T> Responder<'a, T>
+where
+    T: ExchangeHandler,
+{
+    /// Create a new responder.
+    ///
+    /// The `respond_after_ms` parameter instructs the responder how much time to wait before accepting an exchange.
+    ///
+    /// This is useful when utilizing multiple responders on a single `Matter` instance, where e.g. the first (main) responder is the actual one,
+    /// responsible for handling the incoming exchanges, while e.g. another one - with a non-zero `respond_after_ms` - is answerring all exchanges
+    /// not accepted in time by the main responder with a simple "I'm busy, try again later" handling.
+    #[inline(always)]
+    pub const fn new(
+        name: &'a str,
+        handler: T,
+        matter: &'a Matter<'a>,
+        respond_after_ms: u64,
+    ) -> Self {
+        Self {
+            name,
+            handler,
+            matter,
+            respond_after_ms,
+        }
+    }
+
+    /// Get a reference to the `ExchangeHandler` instance used by this responder
+    pub fn handler(&self) -> &T {
+        &self.handler
+    }
+
+    /// Run the responder with a given number of handlers.
+    pub async fn run<const N: usize>(&self) -> Result<(), Error> {
+        info!("{}: Creating {N} handlers", self.name);
+
+        let mut handlers = heapless::Vec::<_, N>::new();
+        info!(
+            "{}: Handlers size: {}B",
+            self.name,
+            core::mem::size_of_val(&handlers)
+        );
+
+        for index in 0..N {
+            let handler_id = (index as u8) + 2;
```

Comment:
the + 2 here could use some documentation or pulled into a constant.
Should we have N be u8 then?

Reply:
Oh that's a left-over from a previous (very different) version version. I'll remove the `+2`.